acl/acl.go

```diff
@@ -603,21 +728,44 @@ func (p *PolicyACL) KeyWrite(key string, scope sentinel.ScopeFn) bool {
 }
 
 // KeyWritePrefix returns if a prefix is allowed to be written
-func (p *PolicyACL) KeyWritePrefix(prefix string) bool {
+func (p *PolicyAuthorizer) KeyWritePrefix(prefix string) bool {
 	// Look for a matching rule that denies
-	_, rule, ok := p.keyRules.LongestPrefix(prefix)
-	if ok && rule.(PolicyRule).aclPolicy != PolicyWrite {
+	prefixAllowed := true
+	found := false
+
+	p.keyRules.WalkPath(prefix, func(path string, leaf interface{}) bool {
+		rule := leaf.(*policyAuthorizerRadixLeaf)
+
+		if rule.prefix != nil {
+			found = true
+			if rule.prefix.(RulePolicy).aclPolicy != PolicyWrite {
+				prefixAllowed = false
+			} else {
+				prefixAllowed = true
+			}
+		}
+		return false
+	})
+
+	if !prefixAllowed {
 		return false
 	}
 
 	// Look if any of our children have a deny policy
 	deny := false
-	p.keyRules.WalkPrefix(prefix, func(path string, rule interface{}) bool {
-		// We have a rule to prevent a write in a sub-directory!
-		if rule.(PolicyRule).aclPolicy != PolicyWrite {
+	p.keyRules.WalkPrefix(prefix, func(path string, leaf interface{}) bool {
```

Comment:
I realise this is the same behaviour as before but it took my ages to work out why this matters. I _think_ the implication is that we need to be able to write the _whole_ prefix so if there is any child we aren't allowed to write then we don't get to write the whole prefix.

Maybe a note in the doc comment pointing out that this is used for example to delete an entire prefix so permission must be granted to write _everything_ under that prefix.